<!--
 * @Author: lu
 * @Date: 2020-09-15 14:55:11
 * @LastEditTime: 2020-09-16 15:02:56
 * @FilePath: \Promise\08_自定义Promise.html
 * @Description: 
-->
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>08_自定义Promise</title>
</head>

<body>
    <script src="./lib/Promise_class.js"></script>
    <!-- <script>
        const p = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve(1)
                // reject(2)

            }, 100)
        })
        p.then(
            value => {
                console.log('onResolved1()', value);
            },
            reason => {
                console.log('onRejected()', reason);
            }
        )
    </script> -->
    <!-- <script>
        new Promise((resolve, reject) => {
            setTimeout(() => {
                // resolve(1)
                reject(2)

            }, 100)
        }).then(
            value => {
                console.log('onResolved1()', value);
            },
            reason => {
                console.log('onRejected1()', reason);
                throw 5
            }
        ).then(
            value => {
                console.log('onResolved2()', value);
            },
            reason => {
                console.log('onRejecte2()', reason);
                throw 5
            }
        ).catch(
            reason => {
                window.console.log('onRejecte4()', reason);
                return new Promise(() => { }) // 中断promise链
            }
        ).then(
            value => {
                console.log('onResolved3()', value);
            },
            reason => {
                console.log('onRejecte3()', reason);
            }
        )
    </script> -->
    <script>
        const p1 = Promise.resolve(2)   // 如果是一般值，p1成功，value就是这个值
        const p2 = Promise.resolve(Promise.resolve(3))  // 如果是成功的promise，p2成功，value是这个promise的value
        const p3 = Promise.resolve(Promise.reject(4))   // 如果是失败的promise，p3失败，reason是这个promise的reason
        p1.then(value => { console.log('p1', value); })
        p2.then(value => { console.log('p2', value); })
        p3.catch(reason => { console.log('p3', reason); })

        const pAll = Promise.all([p1, 7, p2])
        pAll.then(
            value => {
                console.log('all onResolved()', value);
            },
            reason => {
                console.log('all onRejected()', reason);
            }
        )

        const pRace = Promise.race([p3, p2, p1])
        pRace.then(
            value => {
                console.log('race onResolved()', value);
            },
            reason => {
                console.log('race onRejected()', reason);
            }
        )

        const p6 = Promise.resolveDelay(66, 1000)
        p6.then(
            value => {
                console.log('resolveDelay', value);
            }
        )

    </script>
</body>

</html>